### PR TITLE
[keyspace consumer]: Add implementation of keyspace consumer.

### PR DESCRIPTION
**Pull request by oleksandrivantsiv**

### CODE REVIEW

**Comment by oleksandrivantsiv** [comment]
Thanks @JipanYanga for his PR https://github.com/Azure/sonic-swss-common/pull/122 based on which those changes were made.

**Comment by JipanYanga** [comment]
Great progress, thanks for the update!

File: common/schema.h [hunk 1]
```diff
@@ -19,9 +19,7 @@ namespace swss {
 #define APP_NEIGH_TABLE_NAME            "NEIGH_TABLE"
 #define APP_ROUTE_TABLE_NAME            "ROUTE_TABLE"
 #define APP_TUNNEL_DECAP_TABLE_NAME     "TUNNEL_DECAP_TABLE"
-#define APP_MIRROR_SESSION_TABLE_NAME   "MIRROR_SESSION_TABLE"
```
Comment (lguohan) [inline common/schema.h]: suggest to make this a separate PR.

Comment (oleksandrivantsiv) [reply common/schema.h]: Done

File: common/dbconnector.cpp [hunk 2]
```diff
@@ -63,6 +64,22 @@ DBConnector::DBConnector(int db, string unixPath, unsigned int timeout) :
     select(this);
 }
 
+/* No timeout argument, non-blocking DB connection */
+DBConnector::DBConnector(int db, string unixPath) :
+    m_db(db)
+{
+    m_conn = redisConnectUnixNonBlock(unixPath.c_str());
```
Comment (lguohan) [inline common/dbconnector.cpp]: hi sasha, we do not need to have non-block DB connection here. 

if you look at how selectable works in select::select() function, in line 79, we call select(), once the db connect has data, then we call readme() to get the data. since the select guarantees there is data, so the redisGetReply in the readme() will not be blocked. 

once we we get the reply, then we process the reply to get keyspace notification, and then we store in the queue, and then in pops() function, it will read the queue data and then use the get_values() to get the data.

File: common/consumerkeyspacetable.cpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+#include "table.h"
+#include "selectable.h"
+#include "redisselect.h"
+#include "redisapi.h"
+#include "tokenize.h"
+#include "consumerkeyspacetable.h"
+
+namespace swss {
+
+ConsumerKeySpaceTable::ConsumerKeySpaceTable(DBConnector *db, std::string tableName)
```
Comment (lguohan) [inline common/consumerkeyspacetable.cpp]: keyspace is the implementation details, we should rename the table to MultiConsumerTable since the key character is you can have multiple consumer to subscribe the same table and they all get notifications.

File: common/consumerkeyspacetable.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    RedisReply r(m_db, hgetall_key, REDIS_REPLY_ARRAY);
+    redisReply *reply = r.getContext();
+    values.clear();
+
+    if (!reply->elements)
+        return false;
+
+    if (reply->elements & 1)
+        throw std::system_error(make_error_code(std::errc::result_out_of_range),
+                           "Invalid redis reply");
```
Comment (lguohan) [inline common/consumerkeyspacetable.cpp]: what is & 1? what is this error?

**Comment by lguohan** [comment]
for all swss-common code, google unit test is needed. 

for this consumer table, we need test various cases

del/set
del/set/del
set/del/set

set/del/set/del/set/del....

File: common/consumerkeyspacetable.cpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+    for (unsigned int i = 0; i < reply->elements; ++i)
+    {
+        table_entries.push_back(reply->element[i]->str);
+    }
+
+    return true;
+}
+
+bool ConsumerKeySpaceTable::getValues(std::string key, std::vector<FieldValueTuple> &values)
```
Comment (JipanYanga) [inline common/consumerkeyspacetable.cpp]: The implementation is almost the same as bool Table::get(string key, vector<FieldValueTuple> &values).
What is the reason to implement it here again,   "|" table separator?

File: common/consumerkeyspacetable.cpp [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+#include <limits>
+#include <hiredis/hiredis.h>
+#include "dbconnector.h"
+#include "table.h"
+#include "selectable.h"
+#include "redisselect.h"
+#include "redisapi.h"
+#include "tokenize.h"
+#include "consumerkeyspacetable.h"
+
```
Comment (stcheng) [inline common/consumerkeyspacetable.cpp]: could you add `using namespace std` in all `*.cpp` files and remove the corresponding `std::`?

File: common/consumerkeyspacetable.cpp [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        auto ctx = ctx0->element[1];
+        if (m_keyspace.compare(ctx->str))
+        {
+            SWSS_LOG_ERROR("invalid pattern %s returned for pmessage of %s", ctx->str, m_keyspace.c_str());
+            continue;
+        }
+
+        ctx = ctx0->element[2];
+        std::string msg(ctx->str);
+        size_t pos = msg.find(":");
```
Comment (stcheng) [inline common/consumerkeyspacetable.cpp]: could you use the `tokenize` function?

**Comment by oleksandrivantsiv** [comment]
@lguohan Unit tests will be implemented soon.

File: common/consumerkeyspacetable.cpp (hunk 3 above)
Comment (oleksandrivantsiv) [reply common/consumerkeyspacetable.cpp]: Done

File: common/consumerkeyspacetable.cpp (hunk 4 above)
Comment (oleksandrivantsiv) [reply common/consumerkeyspacetable.cpp]: This code is removed from class at all. However, it was taken from: https://github.com/Azure/sonic-swss-common/blob/master/common/table.cpp#L31

File: common/consumerkeyspacetable.cpp (hunk 7 above)
Comment (oleksandrivantsiv) [reply common/consumerkeyspacetable.cpp]: In this case simple find is more comfortable.

File: common/consumerkeyspacetable.cpp (hunk 5 above)
Comment (oleksandrivantsiv) [reply common/consumerkeyspacetable.cpp]: Didn't know that there is an implementation of the same functionality. Thanks for the suggestion.

File: common/consumerkeyspacetable.cpp (hunk 6 above)
Comment (oleksandrivantsiv) [reply common/consumerkeyspacetable.cpp]: done

**Comment by lguohan** [comment]
@qiluo-msft , please review.

**Comment by lguohan** [comment]
@oleksandrivantsiv , can you add unit test to this new class?

File: common/table.cpp [hunk 8]
```diff
@@ -33,8 +33,10 @@ bool Table::get(string key, vector<FieldValueTuple> &values)
                            "Unable to connect netlink socket");
 
     for (unsigned int i = 0; i < reply->elements; i += 2)
-        values.push_back(make_tuple(reply->element[i]->str,
+    {
+        values.push_back(make_tuple(stripSpecialSym(reply->element[i]->str),
```
Comment (qiluo-msft) [inline common/table.cpp]: >stripSpecialSym [](start = 36, length = 15)

Could you add some comments what it's input like, what is the expected action on it? It's hard to understand now.

File: common/multiconsumerstatetable.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+    return Selectable::NODATA;
+}
+
+void MultiConsumerStateTable::readMe()
+{
+    redisReply *reply = NULL;
+
+    if (redisGetReply(m_subscribe->getContext(), (void**)&reply) != REDIS_OK)
+        throw "Unable to read redis reply";
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.cpp]: >throw [](start = 8, length = 5)

throw a runtime_exception, or other specific exception type.

File: common/multiconsumerstatetable.cpp [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+    while (auto ctx0 = getReply())
+    {
+        /* Use RedisReply as auto_ptr to destroy reply object */
+        RedisReply r(ctx0);
+        KeyOpFieldsValuesTuple kco;
+        /* if the Key-space notification is empty, try next one. */
+        if (ctx0->type == REDIS_REPLY_NIL)
+        {
+            continue;
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.cpp]: >continue [](start = 12, length = 8)

Is it burning CPU cycles in a while loop to wait a reply? If yes, we should rethink about the design.

File: common/multiconsumerstatetable.cpp [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+{
+    redisReply *reply = NULL;
+
+    if (redisGetReply(m_subscribe->getContext(), (void**)&reply) != REDIS_OK)
+        throw "Unable to read redis reply";
+
+    m_reply_buffer.push_back(reply);
+}
+
+void MultiConsumerStateTable::pops(deque<KeyOpFieldsValuesTuple> &vkco, string /*prefix*/)
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.cpp]: >vkco [](start = 66, length = 4)

clear vkco first?

File: common/multiconsumerstatetable.cpp [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+{
+    SWSS_LOG_ERROR("MultiConsumerStateTable::pops enter %s", getTableName().c_str());
+
+    if (!m_buffer.empty())
+    {
+        vkco = m_buffer;
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.cpp]: >vkco = m_buffer [](start = 8, length = 15)

vkco.insert(m_buffer.begin(), m_buffer.end())

File: common/multiconsumerstatetable.cpp [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        m_buffer.clear();
+        return;
+    }
+
+    while (auto ctx0 = getReply())
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.cpp]: >ctx0 [](start = 16, length = 4)

ctx0 -> reply

File: common/multiconsumerstatetable.cpp [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    if (!m_buffer.empty() || !m_reply_buffer.empty())
+    {
+        return Selectable::DATA;
+    }
+
+    if (redisGetReplyFromReader(m_subscribe->getContext(),
+                                (void**)&reply) != REDIS_OK)
+    {
+        return Selectable::ERROR;
+    } else if (reply != NULL)
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.cpp]: >else [](start = 6, length = 4)

keep if-else style consistent.

File: common/multiconsumerstatetable.h [hunk 15]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <string>
+#include <deque>
+#include "dbconnector.h"
+#include "consumertablebase.h"
+
+namespace swss {
+
+class MultiConsumerStateTable : public ConsumerTableBase
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.h]: >MultiConsumerStateTable [](start = 6, length = 23)

Name is confusing. It has nothing to do with ConsumerStateTable or Multi.

File: common/multiconsumerstatetable.cpp [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+void MultiConsumerStateTable::readMe()
+{
+    redisReply *reply = NULL;
+
+    if (redisGetReply(m_subscribe->getContext(), (void**)&reply) != REDIS_OK)
+        throw "Unable to read redis reply";
+
+    m_reply_buffer.push_back(reply);
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.cpp]: >reply [](start = 29, length = 5)

How about capturing it by RedisReply wrapper, and storing it in a container, so there is no chance of resource leak?

File: common/multiconsumerstatetable.cpp [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+#include "tokenize.h"
+
+using namespace std;
+
+namespace swss {
+
+MultiConsumerStateTable::MultiConsumerStateTable(DBConnector *db, string tableName)
+    : ConsumerTableBase(db, tableName, "|"), m_table(db, tableName, getSeparator())
+{
+    m_keyspace = "__keyspace@";
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.cpp]: Suggest add some comments at the file header, and unit test, to help others read the code.

File: common/multiconsumerstatetable.h (hunk 15 above)
Comment (JipanYanga) [reply common/multiconsumerstatetable.h]: Maybe we could get some hint from NotificationConsumer, which is quite similar to the keyspace consumer implementation.

File: common/multiconsumerstatetable.cpp [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+    m_keyspace += to_string(db->getDB()) + "__:" + tableName + getSeparator() +"*";
+
+    psubscribe(m_db, m_keyspace);
+
+    vector<string> keys;
+    m_table.getTableKeys(keys);
```
Comment (JipanYanga) [inline common/multiconsumerstatetable.cpp]: In future implementation, we have the need to get some switch level configuration like switch mac, switch flooding control from configDB, and apply the setting to port, LAG, VLAN and etc.   For settings like mac, It is not easy to dynamically modifying it on port/lag/VLAN, so we want to fetch and save them before all keyspace notification handling in the Select framework.   For other settings in switch table, they may be processed as usual in select framework.

Is it possible to provide option for specifying whether the existing keys should be fetched after psubscribe in this constructor?

File: common/multiconsumerstatetable.h (hunk 15 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.h]: Name was suggested by @lguohan:
https://github.com/Azure/sonic-swss-common/pull/129#discussion-diff-141996254R13

Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.h]: @lguohan Do you agree to change the name?

File: common/multiconsumerstatetable.cpp (hunk 9 above)
Comment (lguohan) [inline common/multiconsumerstatetable.cpp]: >throw "Unable to read redis reply"; [](start = 8, length = 35)

add {}

File: common/multiconsumerstatetable.cpp (hunk 16 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: RedisReply doesn't have copy constructor implemented. It is not possible to do without modification of RedisReply class. In pops function every reply pointer is put into RedisReply object to free resourses correctly:
https://github.com/Azure/sonic-swss-common/pull/129/files#diff-e21c43ea3f32ce9e0342a44a352e55bbR91

File: common/multiconsumerstatetable.cpp (hunk 12 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: From the performance point of view it is better to use copy constructor.

File: common/multiconsumerstatetable.cpp (hunk 10 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: No, we just iterating over replies that we got and do validation. Replies are received only after modification of subscribed table.

File: common/table.cpp (hunk 8 above)
Comment (oleksandrivantsiv) [reply common/table.cpp]: sure

File: common/multiconsumerstatetable.cpp (hunk 17 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: sure.

File: common/multiconsumerstatetable.cpp (hunk 14 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: will change the code

File: common/multiconsumerstatetable.cpp (hunk 9 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: sure

File: common/multiconsumerstatetable.cpp (hunk 11 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: agree

File: common/multiconsumerstatetable.cpp (hunk 10 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: In general pops method is called only after event is already received.

Comment (lguohan) [reply common/multiconsumerstatetable.cpp]: why not return? 

if there is no notification, the orchagent will be blocked at
select here.
[https://github.com/Azure/sonic-swss/blob/535458432b8bde67829339c8824ac10fa347bc8f/orchagent/orchdaemon.cpp#L135](https://github.com/Azure/sonic-swss/blob/535458432b8bde67829339c8824ac10fa347bc8f/orchagent/orchdaemon.cpp#L135)


only when there is notification, the orchagent will use pops() to get the data.

[https://github.com/Azure/sonic-swss/blob/6334d57dbd121c52a920473493fe2d927f439bd4/orchagent/orch.cpp#L87](https://github.com/Azure/sonic-swss/blob/6334d57dbd121c52a920473493fe2d927f439bd4/orchagent/orch.cpp#L87)



---
In reply to: [143390924](https://github.com/Azure/sonic-swss-common/pull/129#discussion_r143390924) [](ancestors = 143390924)

Comment (lguohan) [reply common/multiconsumerstatetable.cpp]: see my other comments.

---
In reply to: [143460751](https://github.com/Azure/sonic-swss-common/pull/129#discussion_r143460751) [](ancestors = 143460751)

File: common/multiconsumerstatetable.h (hunk 15 above)
Comment (lguohan) [reply common/multiconsumerstatetable.h]: yes, multiConsumerStateTable is suggested by me. The reason is that it is the state table, since when you get the data, it is the latest state because redis only store the latest value for a key. In this sense, it is the same as consumerstate table. Multi is added because this table supports Mutiple consumer compared with the original consumerstate table.

also, I think we should have a multProducerStateTable is write the the data later. current, we are just using the plain redis set command to do it. I am not sure this is the most optimized way to do it.

Comment (JipanYanga) [reply common/multiconsumerstatetable.h]: Just curious,  it is a typical  pubsub implementation and it is built on redis pubsub mechanism, why are we so keen to call it *consumer* instead of *subscriber*?  Same for *producer* vs *publisher*. 
https://redis.io/topics/pubsub
https://redis.io/topics/notifications

File: common/multiconsumerstatetable.cpp (hunk 16 above)
Comment (qiluo-msft) [reply common/multiconsumerstatetable.cpp]: If at the end any thing is not popped, then it is leaked.

File: common/multiconsumerstatetable.cpp (hunk 12 above)
Comment (qiluo-msft) [reply common/multiconsumerstatetable.cpp]: It is not right in semantics.

File: common/multiconsumerstatetable.cpp (hunk 10 above)
Comment (oivantsiv) [reply common/multiconsumerstatetable.cpp]: getReply method returns reply message from   cache (from m_replay_buffer deque). It is non blocking. We can't return until messages remains in cache.

Comment (oivantsiv) [reply common/multiconsumerstatetable.cpp]: getReply method returns reply message from   cache (from m_replay_buffer deque). It is non blocking. We can't return until messages remains in cache.

Comment (oivantsiv) [reply common/multiconsumerstatetable.cpp]: getReply method returns reply message from   cache (from m_replay_buffer deque). It is non blocking. We can't return until messages remains in cache.

File: common/multiconsumerstatetable.cpp (hunk 12 above)
Comment (oivantsiv) [reply common/multiconsumerstatetable.cpp]: I'm not sure that I understand correctly. Can you explain what wrong in copy constructor?

File: common/multiconsumerstatetable.cpp (hunk 10 above)
Comment (oivantsiv) [reply common/multiconsumerstatetable.cpp]: Key space event is not persistent. For each DB operation it is generated only once. To make MuliConsumerStateTable working with Select framework I had to override "readCache" and "readMe" methods. Those methods reads keyspace events and put them into event queue. "pops" method works only with data stored in the cache. If cache is empty it is not called.

File: common/multiconsumerstatetable.cpp (hunk 16 above)
Comment (oivantsiv) [reply common/multiconsumerstatetable.cpp]: This can happen only if application stuck. And leaked memory is not the for think in this case :). 
Will add copy constructor to RedisReply. Make will use it as a wrapper.

File: common/multiconsumerstatetable.cpp (hunk 10 above)
Comment (oivantsiv) [reply common/multiconsumerstatetable.cpp]: getReply inplementation:
https://github.com/Azure/sonic-swss-common/pull/129/files#diff-e21c43ea3f32ce9e0342a44a352e55bbR159
readCache implementation:
https://github.com/Azure/sonic-swss-common/pull/129/files#diff-e21c43ea3f32ce9e0342a44a352e55bbR45
readMe implementation:
https://github.com/Azure/sonic-swss-common/pull/129/files#diff-e21c43ea3f32ce9e0342a44a352e55bbR67

File: common/multiconsumerstatetable.cpp (hunk 16 above)
Comment (qiluo-msft) [reply common/multiconsumerstatetable.cpp]: Leaking is not acceptable in any case. Without deep investigation, I just feel a copy constructor is not necessary for this situation. Let me know if it is wrong.

---
In reply to: [143554061](https://github.com/Azure/sonic-swss-common/pull/129#discussion_r143554061) [](ancestors = 143554061)

File: common/multiconsumerstatetable.cpp (hunk 12 above)
Comment (qiluo-msft) [reply common/multiconsumerstatetable.cpp]: I am trying to read your mind that you intend to insert all the elements in m_buffer into vkco, and then clear m_buffer. The implement just as what you intend.

vkco.insert(m_buffer.begin(), m_buffer.end());
m_buffer.clear()

Trust the compiler to optimize base on context. If the compiler doesn't utilize the opportunity, try optimize without sacrifice readability. If finally it is impossible, you want to use this trick, give intensive comment like // TRICK:


---
In reply to: [143551979](https://github.com/Azure/sonic-swss-common/pull/129#discussion_r143551979) [](ancestors = 143551979)

File: common/multiconsumerstatetable.cpp [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+        vkco.push_back(kco);
+    }
+
+    m_reply_buffer.clear();
+
+    return;
+}
+
+redisReply *MultiConsumerStateTable::getReply()
```
Comment (qiluo-msft) [inline common/multiconsumerstatetable.cpp]: >getReply [](start = 37, length = 8)

Actually you mean to popReply?

File: common/multiconsumerstatetable.h (hunk 15 above)
Comment (qiluo-msft) [reply common/multiconsumerstatetable.h]: It is still easy to confuse people because the name is pointing against ConsumerStateTable. How about SubscriberStateTable? Just a quick try.

---
In reply to: [143472274](https://github.com/Azure/sonic-swss-common/pull/129#discussion_r143472274) [](ancestors = 143472274)

File: common/multiconsumerstatetable.cpp (hunk 19 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: Yes, will rename and add comments to make it clear

File: common/table.cpp (hunk 8 above)
Comment (oleksandrivantsiv) [reply common/table.cpp]: done

File: common/multiconsumerstatetable.cpp (hunk 19 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: done

File: common/multiconsumerstatetable.cpp (hunk 9 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: done

File: common/multiconsumerstatetable.cpp (hunk 14 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: done

**Comment by qiluo-msft** [comment]
BTW, could you do us a favor _not_ to force push in a pull request? Leaving the old commits will keep discussion context. We have some internal tools to compare between commits so we can easily locate what you modify incrementally.

**Comment by oleksandrivantsiv** [comment]
@qilimsft Will not do push force in the future.

**Comment by oleksandrivantsiv** [comment]
@qiluo-msft @lguohan Unittests are added. Please review.

**Comment by oleksandrivantsiv** [comment]
@lguohan @qiluo-msft Test stuck because keyspace notification is disabled in test environment. To enable it the following line should be added to /etc/redis/redis.conf file:
https://redis.io/topics/notifications
```
notify-keyspace-events AKE
```
Can you please take a look?

File: common/table.cpp (hunk 8 above)
Comment (qiluo-msft) [reply common/table.cpp]: I cannot find any related comments.

---
In reply to: [143827736](https://github.com/Azure/sonic-swss-common/pull/129#discussion_r143827736) [](ancestors = 143827736)

File: common/multiconsumerstatetable.h (hunk 15 above)
Comment (lguohan) [reply common/multiconsumerstatetable.h]: @qiluo-msft, suggests SubscriberStateTable. I agree subscriber is a more accurate name. Consumer consumes the data, and no other consumer can share the same data.

Comment (lguohan) [reply common/multiconsumerstatetable.h]: Yes, i agree.

---
In reply to: [143455459](https://github.com/Azure/sonic-swss-common/pull/129#discussion_r143455459) [](ancestors = 143455459)

File: common/multiconsumerstatetable.cpp (hunk 18 above)
Comment (lguohan) [reply common/multiconsumerstatetable.cpp]: the dependency issue should be solved by statedb, in this case the port/lag/vlan cfgmgr should not be started until someone configure the switch like mac, flooding controls and write them to the state db.

**Comment by qiluo-msft** [comment]
I fixed. Please go ahead.

---
In reply to: [337655769](https://github.com/Azure/sonic-swss-common/pull/129#issuecomment-337655769) [](ancestors = 337655769)

**Comment by lguohan** [comment]
retest this please

File: common/multiconsumerstatetable.cpp (hunk 12 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: Compiler won't optimize that. It can't optimize, change or remove methods logic. 
Copy constructor knows elements count it needs to copy. It will allocate memory once and won't do re-allocation after that. 

Insert method inserts elements one by one and based on elements count re-allocation of internal structure may be required. To achieve the same performance with insert method in this case vkco.reserve(m_buffer.size()) should be called additionally.

File: common/table.cpp (hunk 8 above)
Comment (oleksandrivantsiv) [reply common/table.cpp]: @qiluo-msft I've added description to into header file:
https://github.com/Azure/sonic-swss-common/pull/129/files#diff-980cf2c21abcb620e52c9c6dddfa684fR146

File: common/multiconsumerstatetable.cpp (hunk 17 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: Added description for each function in the header file.

File: common/table.cpp (hunk 8 above)
Comment (oleksandrivantsiv) [reply common/table.cpp]: Description was added in header file.

File: common/multiconsumerstatetable.h (hunk 15 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.h]: Done

File: common/multiconsumerstatetable.cpp (hunk 16 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: Done

File: common/multiconsumerstatetable.cpp (hunk 12 above)
Comment (oleksandrivantsiv) [reply common/multiconsumerstatetable.cpp]: Done

File: tests/redis_multi_state_ut.cpp [hunk 20]
```diff
@@ -0,0 +1,382 @@
+#include <common/subscriberstatetable.h>
```
Comment (qiluo-msft) [inline tests/redis_multi_state_ut.cpp]: #include "common/subscriberstatetable.h"
at the end

Comment (oleksandrivantsiv) [reply tests/redis_multi_state_ut.cpp]: done

File: common/subscriberstatetable.cpp [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+
+        vkco.push_back(kco);
+    }
+
+    m_keyspace_event_buffer.clear();
+
+    return;
+}
+
+shared_ptr<RedisReply> SubscriberStateTable::popEventBuffer()
```
Comment (qiluo-msft) [inline common/subscriberstatetable.cpp]: >RedisReply [](start = 11, length = 10)

Thanks for the improvement. The shared_ptr seems not necessary. If implementing two functions
```
RedisReply& front();
void pop();
```
as STL containers, the shared_ptr could be removed.

Comment (oleksandrivantsiv) [reply common/subscriberstatetable.cpp]: Can this be changed in separated PR? In other case this one never be closed :)

Comment (qiluo-msft) [reply common/subscriberstatetable.cpp]: I prefer changing it in this PR. This is the only concern from my side.

Comment (oleksandrivantsiv) [reply common/subscriberstatetable.cpp]: Can you describe a little bit more? Do you want to remove popEventBuffer and implement those two methods instead? I will be equivalent to just using m_keyspace_event_buffer without wrapper function. What is your concern in using shared_ptr?

Comment (qiluo-msft) [reply common/subscriberstatetable.cpp]: like this
```
void popEventBuffer();
RedisReply& frontEventBuffer();
```
It's ok to directly use m_keyspace_event_buffer. My arguments are
1. shared_ptr introduces time and space overhead
2. you use shared_ptr for preventing implement copy constructor. However the alternative I proposed (or you proposed) may be much easier, and make code clean.

Comment (oleksandrivantsiv) [reply common/subscriberstatetable.cpp]: I have the following concerns on this:
1. It won't be possible to do without modification of RedisReply. To emplace RedisReply object into deque *move* constructor should be implemented. RedisReply is not covered with unit tests. It will be hard to verify that modification of this class doesn't have side effects.
2. shared_prt is efficient enough. Time and space overhead is minimal.
3. It will have negative effect on code readability because object should reference should be taken in one place and popped in another. With shared_ptr it is enough to get object in one place and it will be deleted automatically.

Comment (oleksandrivantsiv) [reply common/subscriberstatetable.cpp]: An article which describes memory and performance overhead given by smart pointers:
http://www.modernescpp.com/index.php/memory-and-performance-overhead-of-smart-pointer

File: common/subscriberstatetable.cpp [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+    while (auto event = popEventBuffer())
+    {
+        KeyOpFieldsValuesTuple kco;
+        /* if the Key-space notification is empty, try next one. */
+        if (event->getContext()->type == REDIS_REPLY_NIL)
+        {
+            continue;
+        }
+
+        assert(event->type == REDIS_REPLY_ARRAY);
```
Comment (qiluo-msft) [inline common/subscriberstatetable.cpp]: >event->type [](start = 15, length = 11)

Build break under debug mode.

**Comment by lguohan** [comment]
@oleksandrivantsiv , please fix the debug mode build error in next PR. 

Thanks a lot for adding this.
